content/blog/monkey-patch.md

```diff
@@ -0,0 +1,65 @@
++++
+title = "Monkey Patching"
+slug = "monkey-patch"
+date = 2023-10-17T18:29:54+05:30
+image = "/images/2023/10/monkeypatch.jpg"
+draft = false
+authors = ["Nitin Rajkumar"]
+description = "Monkey patching and it's sideeffects in Ruby"
+tags = ["Software Craftsmanship", "GraphQL"]
+categories = ["Software Craftsmanship", "Ruby", "Monkey Patching"]
+type = ""
++++
+
+Ruby, being a dynamic programming language provides the ability and hence freedom to reopen the existing classes and change their behavior during runtime by adding new methods or modifying existing methods. 
+
+#### How to monkey patch
+
+```ruby
+class Sound
+  def car
+    "honnnnnk..."
+  end
+end
+
+class Sound
+  def bike
+    "peep-peep"
+  end
+end
+sound = Sound.new
+sound.car #=> honnnnnk...
+sound.bike #=> peep-peep
+```
+The above is a simple example where we reopened Sound class and added a new method. It provides many other opportunities including overriding the method of the original class.
+
+#### When does one think of using it
+Usually when the programmer wants to fix the bug quickly.
+1. When the gem or library that is being used has a bug and before a patch-fix from the library if one wants to fix it in their code base.
+2. When certain method you expect the library to contain is missing, one can monkey patch the method.
+
+#### Cons
+
+```ruby
+class String
+  def upcase
+    # do nothing 
+  end
+end
+```
+The above monkey patch would return nothing causing the `upcase` to fail everywhere the patch is being used.
+1. In a large code base multiple developers may be monkey patching the same method in different ways.
+2. When a 3rd-party gem is monkey patched, upgrading the gem may result in conflict.
+3. Can be difficult to track which patch is giving the result.
+4. Adds to tech debt as monkey patching needs to be cleaned out once the fix is in place.
```

Comment:
Grammar improvement by Chat GPT

```suggestion
The above monkey patch would return nothing, causing the upcase method to fail wherever the patch is applied.

1. In a large codebase, multiple developers may be monkey patching the same method in different ways.

2. When a third-party gem is monkey patched, upgrading the gem may result in conflicts.

3. It can be difficult to determine which patch is producing the desired result.

4. This adds to technical debt, as monkey patches need to be removed once the fix is in place.
```